File: client-python/zeromq.py
import json

##########################################################

import zmq # pip install pyzmq

##########################################################

zeromq_boolRunning = False

##########################################################

def zeromq_start():
    global zeromq_boolRunning
    
    zeromq_boolRunning = True
    
    objContext = zmq.Context()
    objSocket = objContext.socket(zmq.PAIR); objSocket.bind('tcp://*:' + str(main_intZeromq))
    
    while zeromq_boolRunning == True:
        objIn = json.loads(objSocket.recv().decode('utf-8'))
        objOut = {}
        
        if objIn['strFunction'] == 'ping':
            objOut['strOut'] = main_strName
            
        elif objIn['strFunction'] == 'chess_reset':
            chess_reset()
            
        elif objIn['strFunction'] == 'chess_boardGet':
            objOut['strOut'] = chess_boardGet()
            
        elif objIn['strFunction'] == 'chess_boardSet':
            chess_boardSet(objIn['strIn'])
            
        elif objIn['strFunction'] == 'chess_winner':
            objOut['strReturn'] = chess_winner()
            
        elif objIn['strFunction'] == 'chess_isValid':
            objOut['boolReturn'] = chess_isValid(objIn['intX'], objIn['intY'])
            
        elif objIn['strFunction'] == 'chess_isEnemy':
            objOut['boolReturn'] = chess_isEnemy(objIn['strPiece'])
            
        elif objIn['strFunction'] == 'chess_isOwn':
            objOut['boolReturn'] = chess_isOwn(objIn['strPiece'])
            
        elif objIn['strFunction'] == 'chess_isNothing':
            objOut['boolReturn'] = chess_isNothing(objIn['strPiece'])
            
        elif objIn['strFunction'] == 'chess_eval':
            objOut['intReturn'] = chess_eval()
            
        elif objIn['strFunction'] == 'chess_moves':
            strOut = chess_moves()
            
            objOut['intOut'] = len(strOut)
            objOut['strOut'] = str.join('', strOut)
            
        elif objIn['strFunction'] == 'chess_movesShuffled':
            strOut = chess_movesShuffled()
            
            objOut['intOut'] = len(strOut)
            objOut['strOut'] = str.join('', strOut)
            
        elif objIn['strFunction'] == 'chess_movesEvaluated':
            strOut = chess_movesEvaluated()
            
            objOut['intOut'] = len(strOut)
            objOut['strOut'] = str.join('', strOut)
            
        elif objIn['strFunction'] == 'chess_move':
            chess_move(objIn['strIn'])
            
        elif objIn['strFunction'] == 'chess_moveRandom':
            objOut['strOut'] = chess_moveRandom()
            
        elif objIn['strFunction'] == 'chess_moveGreedy':
            objOut['strOut'] = chess_moveGreedy()
            
        elif objIn['strFunction'] == 'chess_moveNegamax':
            objOut['strOut'] = chess_moveNegamax(objIn['intMaxdepth'], objIn['intDuration'])
            
        elif objIn['strFunction'] == 'chess_moveAlphabeta':
            objOut['strOut'] = chess_moveAlphabeta(objIn['intMaxdepth'], objIn['intDuration'])
            
        elif objIn['strFunction'] == 'chess_undo':
            chess_undo()

        # end
        
        objSocket.send_string(json.dumps(objOut))
    # end
    
    objSocket.close()
    objContext.destroy()
# end

def zeromq_stop():
    global zeromq_boolRunning

    zeromq_boolRunning = false
# end

File: client-java/zeromq.java
import java.util.Vector;

import org.json.JSONObject;
import org.zeromq.ZMQ;

public class zeromq {
    public static boolean boolRunning = false;
    
    public static void start() {
        zeromq.boolRunning = true;

        ZMQ.Context objContext = ZMQ.context(1);
        ZMQ.Socket objSocket = objContext.socket(ZMQ.PAIR); objSocket.bind("tcp://*:" + main.intZeromq);
    
        do {
            JSONObject objIn = new JSONObject(objSocket.recvStr());
            JSONObject objOut = new JSONObject();
        
            if (objIn.getString("strFunction").equals("ping") == true) {
                objOut.put("strOut", main.strName);
                
            } else if (objIn.getString("strFunction").equals("chess_reset") == true) {
                chess.reset();
                
            } else if (objIn.getString("strFunction").equals("chess_boardGet") == true) {
                objOut.put("strOut", chess.boardGet());
                
            } else if (objIn.getString("strFunction").equals("chess_boardSet") == true) {
                chess.boardSet(objIn.getString("strIn"));
                
            } else if (objIn.getString("strFunction").equals("chess_winner") == true) {
                objOut.put("strReturn", chess.winner() + "\0");
                
            } else if (objIn.getString("strFunction").equals("chess_isValid") == true) {
                objOut.put("boolReturn", chess.isValid(objIn.getInt("intX"), objIn.getInt("intY")));
                
            } else if (objIn.getString("strFunction").equals("chess_isEnemy") == true) {
                objOut.put("boolReturn", chess.isEnemy(objIn.getString("strPiece").charAt(0)));
                
            } else if (objIn.getString("strFunction").equals("chess_isOwn") == true) {
                objOut.put("boolReturn", chess.isOwn(objIn.getString("strPiece").charAt(0)));
                
            } else if (objIn.getString("strFunction").equals("chess_isNothing") == true) {
                objOut.put("boolReturn", chess.isNothing(objIn.getString("strPiece").charAt(0)));
                
            } else if (objIn.getString("strFunction").equals("chess_eval") == true) {
                objOut.put("intReturn", chess.eval());
                
            } else if (objIn.getString("strFunction").equals("chess_moves") == true) {
                Vector<String> strOut = chess.moves();
                
                objOut.put("intOut", strOut.size());
                objOut.put("strOut", "");
                
                for (int intMove = 0; intMove < strOut.size(); intMove += 1) {
                    objOut.put("strOut", objOut.get("strOut") + strOut.get(intMove));
                }
                
            } else if (objIn.getString("strFunction").equals("chess_movesShuffled") == true) {
                Vector<String> strOut = chess.movesShuffled();
                
                objOut.put("intOut", strOut.size());
                objOut.put("strOut", "");
                
                for (int intMove = 0; intMove < strOut.size(); intMove += 1) {
                    objOut.put("strOut", objOut.get("strOut") + strOut.get(intMove));
                }
                
            } else if (objIn.getString("strFunction").equals("chess_movesEvaluated") == true) {
                Vector<String> strOut = chess.movesEvaluated();
                
                objOut.put("intOut", strOut.size());
                objOut.put("strOut", "");
                
                for (int intMove = 0; intMove < strOut.size(); intMove += 1) {
                    objOut.put("strOut", objOut.get("strOut") + strOut.get(intMove));
                }
                
            } else if (objIn.getString("strFunction").equals("chess_move") == true) {
                chess.move(objIn.getString("strIn"));
                
            } else if (objIn.getString("strFunction").equals("chess_moveRandom") == true) {
                objOut.put("strOut", chess.moveRandom());
                
            } else if (objIn.getString("strFunction").equals("chess_moveGreedy") == true) {
                objOut.put("strOut", chess.moveGreedy());
                
            } else if (objIn.getString("strFunction").equals("chess_moveNegamax") == true) {
                objOut.put("strOut", chess.moveNegamax(objIn.getInt("intMaxdepth"), objIn.getInt("intDuration")));
                
            } else if (objIn.getString("strFunction").equals("chess_moveAlphabeta") == true) {
                objOut.put("strOut", chess.moveAlphabeta(objIn.getInt("intMaxdepth"), objIn.getInt("intDuration")));
                
            } else if (objIn.getString("strFunction").equals("chess_undo") == true) {
                chess.undo();
                
            }

            objSocket.send(objOut.toString());
        } while (zeromq.boolRunning == true);
        
        objSocket.close();
        objContext.term();
    }
    
    public static void stop() {
        zeromq.boolRunning = false;
    }
}
